dbms/src/Storages/DeltaMerge/Segment.cpp

```diff
@@ -500,51 +501,54 @@ BlockInputStreamPtr Segment::getInputStreamForDataExport(const DMContext & dm_co
 BlockInputStreamPtr Segment::getInputStreamRaw(const DMContext & dm_context,
                                                const ColumnDefines & columns_to_read,
                                                const SegmentSnapshotPtr & segment_snap,
-                                               bool do_range_filter,
+                                               const RowKeyRanges & data_ranges,
+                                               const RSOperatorPtr & filter,
+                                               bool filter_delete_mark,
                                                size_t expected_block_size)
 {
     auto new_columns_to_read = std::make_shared<ColumnDefines>();
 
-    if (!do_range_filter)
+
+    new_columns_to_read->push_back(getExtraHandleColumnDefine(is_common_handle));
+    if (filter_delete_mark)
     {
-        (*new_columns_to_read) = columns_to_read;
+        new_columns_to_read->push_back(getTagColumnDefine());
     }
-    else
-    {
-        new_columns_to_read->push_back(getExtraHandleColumnDefine(is_common_handle));
 
-        for (const auto & c : columns_to_read)
-        {
-            if (c.id != EXTRA_HANDLE_COLUMN_ID)
-                new_columns_to_read->push_back(c);
-        }
+    for (const auto & c : columns_to_read)
+    {
+        if (c.id != EXTRA_HANDLE_COLUMN_ID && (!(filter_delete_mark && c.id == TAG_COLUMN_ID)))
+            new_columns_to_read->push_back(c);
     }
 
-
-    BlockInputStreamPtr delta_stream = std::make_shared<DeltaValueInputStream>(dm_context, //
-                                                                               segment_snap->delta,
-                                                                               new_columns_to_read,
-                                                                               this->rowkey_range);
-
-    RowKeyRanges rowkey_ranges{rowkey_range};
     BlockInputStreamPtr stable_stream = segment_snap->stable->getInputStream(
         dm_context,
         *new_columns_to_read,
-        rowkey_ranges,
-        EMPTY_FILTER,
+        data_ranges,
+        filter,
         std::numeric_limits<UInt64>::max(),
         expected_block_size,
         false);
 
-    if (do_range_filter)
+    BlockInputStreamPtr delta_stream = std::make_shared<DeltaValueInputStream>(dm_context, segment_snap->delta, new_columns_to_read, this->rowkey_range);
+
+    //delta_stream = std::make_shared<SquashingBlockInputStream>(delta_stream, 16384, 0, /*req_id=*/""); // For test
+
+    delta_stream = std::make_shared<DMRowKeyFilterBlockInputStream<false>>(delta_stream, data_ranges, 0);
+    stable_stream = std::make_shared<DMRowKeyFilterBlockInputStream<true>>(stable_stream, data_ranges, 0);
+
+    if (filter_delete_mark)
+    {
+        delta_stream = std::make_shared<DMDeleteFilterBlockInputStream>(delta_stream, columns_to_read);
+        stable_stream = std::make_shared<DMDeleteFilterBlockInputStream>(stable_stream, columns_to_read);
```

Comment:
Add `dm_context.tracing_id` to the logger of `DMDeleteFilterBlockInputStream` for better tracing logs